File: ai-service/src/workflowExpand.ts
import type { StepDraft, WorkflowDraft } from './workflowSchema.js'

type StepSpec = Pick<StepDraft, 'channel' | 'delay_value' | 'delay_unit'>

const CATEGORY_DEFAULTS: Record<string, StepSpec[]> = {
  onboarding: [
    { channel: 'email', delay_value: 0, delay_unit: 'minute' },
    { channel: 'sms', delay_value: 1, delay_unit: 'day' }
  ],
  abandoned_basket: [
    { channel: 'email', delay_value: 1, delay_unit: 'hour' },
    { channel: 'sms', delay_value: 1, delay_unit: 'day' }
  ],
  reactivation: [
    { channel: 'email', delay_value: 0, delay_unit: 'minute' },
    { channel: 'sms', delay_value: 7, delay_unit: 'day' }
  ],
  loyalty: [
    { channel: 'email', delay_value: 0, delay_unit: 'minute' },
    { channel: 'email', delay_value: 3, delay_unit: 'day' }
  ],
  retention: [
    { channel: 'email', delay_value: 0, delay_unit: 'minute' },
    { channel: 'sms', delay_value: 2, delay_unit: 'day' }
  ],
  nurturing: [
    { channel: 'email', delay_value: 0, delay_unit: 'minute' },
    { channel: 'email', delay_value: 2, delay_unit: 'day' }
  ],
  activation: [
    { channel: 'email', delay_value: 0, delay_unit: 'minute' },
    { channel: 'sms', delay_value: 1, delay_unit: 'day' }
  ],
  qualification: [
    { channel: 'email', delay_value: 0, delay_unit: 'minute' },
    { channel: 'email', delay_value: 1, delay_unit: 'day' }
  ]
}

const JOURNEY_RE = /journey|campaign|workflow|series|reminder|recovery|win.?back|hành trình|chuỗi|chiến dịch/i

function detectChannel (text: string): StepDraft['channel'] | null {
  const t = text.toLowerCase()
  if (/\b(email|mail|thư)\b/.test(t)) return 'email'
  if (/\b(sms|text message|tin nhắn)\b/.test(t)) return 'sms'
  if (/\brcs\b/.test(t)) return 'rcs'
  if (/\b(voice|call|gọi)\b/.test(t)) return 'voice'
  return null
}

function normalizeUnit (raw: string): StepDraft['delay_unit'] {
  const u = raw.toLowerCase()
  if (u.startsWith('hour') || u === 'giờ') return 'hour'
  if (u.startsWith('minute') || u === 'phút') return 'minute'
  return 'day'
}

function parseStepCount (prompt: string): number | null {
  const m = prompt.toLowerCase().match(/(\d+)[\s-]*(?:step|steps|bước)/)
  if (!m) return null
  return Math.max(2, Math.min(8, Number(m[1])))
}

function parseExplicitSteps (prompt: string): StepSpec[] {
  const parts = prompt.split(/[,;]|\bthen\b|sau đó|tiếp theo|rồi/i)
  const steps: StepSpec[] = []

  for (const part of parts) {
    const channel = detectChannel(part)
    if (!channel) continue

    let delay_value = 0
    let delay_unit: StepDraft['delay_unit'] = 'minute'

    if (/immediately|right away|ngay|lập tức/i.test(part)) {
      delay_value = 0
      delay_unit = 'minute'
    } else {
      const after = part.match(
        /(?:after|sau)\s+(\d+)\s*(hour|hours|day|days|minute|minutes|giờ|ngày|phút)/i
      )
      const inline = part.match(/(\d+)\s*(hour|hours|day|days|minute|minutes|giờ|ngày|phút)/i)
      const match = after ?? inline
      if (match) {
        delay_value = Number(match[1])
        delay_unit = normalizeUnit(match[2]!)
      }
    }

    steps.push({ channel, delay_value, delay_unit })
  }

  return steps
}

function buildTargetSteps (prompt: string, category: WorkflowDraft['category']): StepSpec[] | null {
  const explicit = parseExplicitSteps(prompt)
  if (explicit.length >= 2) return explicit

  const requested = parseStepCount(prompt)
  if (requested) {
    const channels: StepDraft['channel'][] = []
    for (const part of prompt.split(/[,;]|\band\b|và/i)) {
      const ch = detectChannel(part)
      if (ch && !channels.includes(ch)) channels.push(ch)
    }
    const pool = channels.length > 0 ? channels : (['email', 'sms'] as const)
    return Array.from({ length: requested }, (_, i) => ({
      channel: pool[i % pool.length]!,
      delay_value: i === 0 ? 0 : i,
      delay_unit: (i === 0 ? 'minute' : 'day') as StepDraft['delay_unit']
    }))
  }

  if (JOURNEY_RE.test(prompt)) {
    return CATEGORY_DEFAULTS[category] ?? null
  }

  if (explicit.length === 1) {
    const defaults = CATEGORY_DEFAULTS[category]
    const first = explicit[0]!
    let follow = defaults?.[1] ?? { channel: 'sms' as const, delay_value: 1, delay_unit: 'day' as const }
    if (first.channel === follow.channel) {
      follow = first.channel === 'email'
        ? { channel: 'sms', delay_value: 1, delay_unit: 'day' }
        : { channel: 'email', delay_value: 1, delay_unit: 'day' }
    }
    return [first, follow]
  }

  return null
}

function makeTemplate (
  channel: StepDraft['channel'],
  category: WorkflowDraft['category'],
  stepIndex: number,
  locale: string,
  existing?: StepDraft['template']
): StepDraft['template'] {
  if (existing?.body) {
    return {
      name: existing.name || `Step ${stepIndex + 1}`,
      subject: existing.subject ?? '',
      body: existing.body
    }
  }

  const vi = locale === 'vi'
  if (channel === 'email') {
    const subjects: Record<string, string[]> = {
      onboarding: vi ? ['Chào mừng!', 'Bắt đầu ngay'] : ['Welcome aboard!', 'Get started with us'],
      abandoned_basket: vi ? ['Giỏ hàng đang chờ', 'Hoàn tất đơn'] : ['You left something behind', 'Complete your order'],
      reactivation: vi ? ['Chúng tôi nhớ bạn!', 'Quay lại nhận ưu đãi'] : ['We miss you!', 'Come back for an offer']
    }
    const pool = subjects[category] ?? subjects.onboarding!
    return {
      name: vi ? `Email bước ${stepIndex + 1}` : `Email step ${stepIndex + 1}`,
      subject: pool[stepIndex % pool.length]!,
      body: vi ? '<p>Cảm ơn bạn đã đồng hành.</p>' : '<p>Thanks for being with us.</p>'
    }
  }

  return {
    name: vi ? `SMS bước ${stepIndex + 1}` : `SMS step ${stepIndex + 1}`,
    subject: '',
    body: vi ? 'Cập nhật dành cho bạn.' : 'Your update is ready.'
  }
}

export function expandWorkflowSteps (
  workflow: WorkflowDraft,
  prompt: string,
  locale = 'en'
): WorkflowDraft {
  const steps = workflow.steps ?? []
  let target = buildTargetSteps(prompt, workflow.category)

  if (!target) {
    if (steps.length >= 2) return workflow
    target = CATEGORY_DEFAULTS[workflow.category] ?? null
    if (!target || steps.length >= target.length) return workflow
  }

  if (steps.length >= target.length) return workflow

  const expanded: StepDraft[] = target.map((spec, i) => {
    const existing = steps[i]
    if (existing) {
      const template = makeTemplate(
        spec.channel,
        workflow.category,
        i,
        locale,
        existing.template
      )
      return {
        ...existing,
        channel: spec.channel,
        delay_value: spec.delay_value,
        delay_unit: spec.delay_unit,
        template: existing.template?.body ? existing.template : template,
        ...(spec.channel === 'email'
          ? { email_subject: existing.email_subject ?? template.subject }
          : {})
      }
    }

    const template = makeTemplate(spec.channel, workflow.category, i, locale)
    const step: StepDraft = {
      channel: spec.channel,
      delay_value: spec.delay_value,
      delay_unit: spec.delay_unit,
      template
    }
    if (spec.channel === 'email') {
      step.email_subject = template.subject
    }
    return step
  })

  return { ...workflow, steps: expanded }
}

File: ml/inference/workflow_expand.py
"""Expand workflow steps when model output is shorter than the prompt implies."""

from __future__ import annotations

import re
from typing import Any

UNIT_MAP = {
    "hour": "hour",
    "hours": "hour",
    "giờ": "hour",
    "day": "day",
    "days": "day",
    "ngày": "day",
    "minute": "minute",
    "minutes": "minute",
    "phút": "minute",
}

CATEGORY_DEFAULTS: dict[str, list[dict[str, Any]]] = {
    "onboarding": [
        {"channel": "email", "delay_value": 0, "delay_unit": "minute"},
        {"channel": "sms", "delay_value": 1, "delay_unit": "day"},
    ],
    "abandoned_basket": [
        {"channel": "email", "delay_value": 1, "delay_unit": "hour"},
        {"channel": "sms", "delay_value": 1, "delay_unit": "day"},
    ],
    "reactivation": [
        {"channel": "email", "delay_value": 0, "delay_unit": "minute"},
        {"channel": "sms", "delay_value": 7, "delay_unit": "day"},
    ],
    "loyalty": [
        {"channel": "email", "delay_value": 0, "delay_unit": "minute"},
        {"channel": "email", "delay_value": 3, "delay_unit": "day"},
    ],
    "retention": [
        {"channel": "email", "delay_value": 0, "delay_unit": "minute"},
        {"channel": "sms", "delay_value": 2, "delay_unit": "day"},
    ],
    "nurturing": [
        {"channel": "email", "delay_value": 0, "delay_unit": "minute"},
        {"channel": "email", "delay_value": 2, "delay_unit": "day"},
    ],
    "activation": [
        {"channel": "email", "delay_value": 0, "delay_unit": "minute"},
        {"channel": "sms", "delay_value": 1, "delay_unit": "day"},
    ],
    "qualification": [
        {"channel": "email", "delay_value": 0, "delay_unit": "minute"},
        {"channel": "email", "delay_value": 1, "delay_unit": "day"},
    ],
}

JOURNEY_KEYWORDS = re.compile(
    r"journey|campaign|workflow|series|reminder|recovery|win.?back|hành trình|chuỗi|chiến dịch",
    re.I,
)


def _normalize_unit(raw: str) -> str:
    return UNIT_MAP.get(raw.lower(), "day")


def _detect_channel(text: str) -> str | None:
    t = text.lower()
    if re.search(r"\b(email|mail|thư)\b", t):
        return "email"
    if re.search(r"\b(sms|text message|tin nhắn)\b", t):
        return "sms"
    if re.search(r"\brcs\b", t):
        return "rcs"
    if re.search(r"\b(voice|call|gọi)\b", t):
        return "voice"
    return None


def parse_step_count(prompt: str) -> int | None:
    p = prompt.lower()
    m = re.search(r"(\d+)[\s-]*(?:step|steps|bước)", p)
    if m:
        return max(2, min(8, int(m.group(1))))
    return None


def parse_explicit_steps(prompt: str) -> list[dict[str, Any]]:
    """Parse comma/then-separated channel + delay clauses from the prompt."""
    parts = re.split(r"[,;]|\bthen\b|sau đó|tiếp theo|rồi", prompt, flags=re.I)
    steps: list[dict[str, Any]] = []

    for part in parts:
        channel = _detect_channel(part)
        if not channel:
            continue

        delay_value, delay_unit = 0, "minute"
        if re.search(r"immediately|right away|ngay|lập tức", part, re.I):
            delay_value, delay_unit = 0, "minute"
        else:
            m = re.search(
                r"(?:after|sau)\s+(\d+)\s*(hour|hours|day|days|minute|minutes|giờ|ngày|phút)",
                part,
                re.I,
            )
            if m:
                delay_value = int(m.group(1))
                delay_unit = _normalize_unit(m.group(2))
            else:
                m2 = re.search(
                    r"(\d+)\s*(hour|hours|day|days|minute|minutes|giờ|ngày|phút)",
                    part,
                    re.I,
                )
                if m2:
                    delay_value = int(m2.group(1))
                    delay_unit = _normalize_unit(m2.group(2))

        steps.append({"channel": channel, "delay_value": delay_value, "delay_unit": delay_unit})

    return steps


def build_target_steps(prompt: str, category: str) -> list[dict[str, Any]] | None:
    explicit = parse_explicit_steps(prompt)
    if len(explicit) >= 2:
        return explicit

    requested = parse_step_count(prompt)
    if requested:
        channels: list[str] = []
        for part in re.split(r"[,;]|\band\b|và", prompt, flags=re.I):
            ch = _detect_channel(part)
            if ch and ch not in channels:
                channels.append(ch)
        if not channels:
            channels = ["email", "sms"]

        steps: list[dict[str, Any]] = []
        for i in range(requested):
            steps.append(
                {
                    "channel": channels[i % len(channels)],
                    "delay_value": 0 if i == 0 else i,
                    "delay_unit": "minute" if i == 0 else "day",
                }
            )
        return steps

    if JOURNEY_KEYWORDS.search(prompt):
        defaults = CATEGORY_DEFAULTS.get(category)
        if defaults:
            return [dict(s) for s in defaults]

    defaults = CATEGORY_DEFAULTS.get(category)
    if defaults and len(explicit) == 1:
        # Prompt mentions one channel but implies a journey — add category default follow-up
        first = explicit[0]
        follow = defaults[1] if len(defaults) > 1 else {"channel": "sms", "delay_value": 1, "delay_unit": "day"}
        if first["channel"] == follow["channel"]:
            follow = {"channel": "sms" if first["channel"] == "email" else "email", "delay_value": 1, "delay_unit": "day"}
        return [first, follow]

    return None


def _make_template(
    channel: str,
    category: str,
    step_index: int,
    locale: str,
    existing: dict[str, Any] | None = None,
) -> dict[str, Any]:
    if existing and existing.get("body"):
        return {
            "name": existing.get("name") or f"Step {step_index + 1}",
            "subject": existing.get("subject", ""),
            "body": existing.get("body", ""),
        }

    vi = locale == "vi"
    if channel == "email":
        subjects = {
            "onboarding": ["Welcome aboard!", "Get started with us"] if not vi else ["Chào mừng!", "Bắt đầu ngay"],
            "abandoned_basket": ["You left something behind", "Complete your order"]
            if not vi
            else ["Giỏ hàng đang chờ", "Hoàn tất đơn hàng"],
            "reactivation": ["We miss you!", "Come back for an offer"]
            if not vi
            else ["Chúng tôi nhớ bạn!", "Quay lại nhận ưu đãi"],
        }
        pool = subjects.get(category, subjects["onboarding"])
        return {
            "name": f"Email step {step_index + 1}" if not vi else f"Email bước {step_index + 1}",
            "subject": pool[step_index % len(pool)],
            "body": "<p>Thanks for being with us.</p>" if not vi else "<p>Cảm ơn bạn đã đồng hành.</p>",
        }

    return {
        "name": f"SMS step {step_index + 1}" if not vi else f"SMS bước {step_index + 1}",
        "subject": "",
        "body": "Your update is ready." if not vi else "Cập nhật dành cho bạn.",
    }


def expand_workflow(workflow: dict[str, Any], prompt: str, locale: str = "en") -> dict[str, Any]:
    """Ensure workflow has enough steps based on prompt intent and category defaults."""
    steps = workflow.get("steps") or []
    if not isinstance(steps, list):
        steps = []

    category = workflow.get("category") or "onboarding"
    target = build_target_steps(prompt, category)

    if not target:
        if len(steps) >= 2:
            return workflow
        target = CATEGORY_DEFAULTS.get(category)
        if not target or len(steps) >= len(target):
            return workflow

    if len(steps) >= len(target):
        return workflow

    expanded: list[dict[str, Any]] = []
    for i, spec in enumerate(target):
        if i < len(steps):
            step = dict(steps[i])
            step["channel"] = spec["channel"]
            step["delay_value"] = spec["delay_value"]
            step["delay_unit"] = spec["delay_unit"]
        else:
            tpl = _make_template(spec["channel"], category, i, locale)
            step = {
                "channel": spec["channel"],
                "delay_value": spec["delay_value"],
                "delay_unit": spec["delay_unit"],
                "template": tpl,
            }
            if spec["channel"] == "email":
                step["email_subject"] = tpl["subject"]

        expanded.append(step)

    workflow = dict(workflow)
    workflow["steps"] = expanded
    return workflow
